src/freenet/client/filter/TheoraPacketFilter.java

```diff
@@ -1,148 +1,287 @@
-
-					/* This is a value from an enumerated list of the available color spaces, given in Table.
-					 * The 'Undefined' value indicates that color space information was not available to the encoder.
-					 * It MAY be specified by the application via an external means.
-					 * If a 'Reserved' value is given, a decoder MAY refuse to decode the stream.
-					 * Value Color Space
-					 *  0     Undefined.
-					 *  1     Rec. 470M.
-					 *  2     Rec. 470BG.
-					 *  3     Reserved.
-					 * https://www.theora.org/doc/Theora.pdf CHAPTER 6. BITSTREAM HEADERS page 44 */
-					if(!(CS == 0 || CS == 1 || CS == 2)) throw new UnknownContentTypeException("Header CS: " + CS);
-
-					if(PF == 1) throw new UnknownContentTypeException("Header PF: " + PF);
-					if(Res != 0) throw new UnknownContentTypeException("Header Res: " + Res);
-
-					currentState = State.IDENTIFICATION_FOUND;
+					checkMagicHeader(magicHeader, (byte) 0x80); // -128
+
+					int VMAJ = input.readInt(8);
+					if (VMAJ != 3) throw new UnknownContentTypeException("Header VMAJ: " + VMAJ);
+
+					int VMIN = input.readInt(8);
+					if (VMIN != 2) throw new UnknownContentTypeException("Header VMIN: " + VMIN);
+
+					int VREV = input.readInt(8);
+					if (VREV > 1) throw new UnknownContentTypeException("Header VREV: " + VREV);
+
+					int FMBW = input.readInt(16);
+					if (FMBW == 0) throw new UnknownContentTypeException("Header FMBW: " + FMBW);
+
+					int FMBH = input.readInt(16);
+					if (FMBH == 0) throw new UnknownContentTypeException("Header FMBH: " + FMBH);
+
+					int PICW = input.readInt(24);
+					if (PICW > FMBW*16) throw new UnknownContentTypeException("Header PICW: " + PICW + "; FMBW: " + FMBW);
+
+					int PICH = input.readInt(24);
+					if (PICH > FMBH*16) throw new UnknownContentTypeException("Header PICH: " + PICH + "; FMBH: " + FMBH);
+
+					int PICX = input.readInt(8);
+					if (PICX > FMBW*16-PICX) throw new UnknownContentTypeException("Header PICX: " + PICX + "; FMBW: " + FMBW + "; PICX: " + PICX);
+
+					int PICY = input.readInt(8);
+					if (PICY > FMBH*16-PICY) throw new UnknownContentTypeException("Header PICY: " + PICY + "; FMBH: " + FMBH + "; PICY: " + PICY);
+
+					int FRN = input.readInt(32);
+					if (FRN == 0) throw new UnknownContentTypeException("Header FRN: " + FRN);
+
+					int FRD = input.readInt(32);
+					if (FRD == 0) throw new UnknownContentTypeException("Header FRN: " + FRN);
+
+					int PARN = input.readInt(24);
+					int PARD = input.readInt(24);
+
+					int CS = input.readInt(8);
+					if (!(CS == 0 || CS == 1 || CS == 2)) throw new UnknownContentTypeException("Header CS: " + CS);
+
+					int NOMBR = input.readInt(24);
+					int QUAL = input.readInt(6);
+					int KFGSHIFT = input.readInt(5);
+
+					int PF = input.readInt(2);
+					if (PF == 1) throw new UnknownContentTypeException("Header PF: " + PF);
+
+					int Res = input.readInt(3);
+					if (Res != 0) throw new UnknownContentTypeException("Header Res: " + Res);
+
```

Comment:
Now I skip ignored values. Should I think about some other streamlining?

Reply:
I totally would write a method like this:

```
void checkHeaderField(String name, int sizeInBits, Predicate<Int> validator) {
    int value = input.readInt(sizeInBits);
    if (!validator.apply(value)) {
        throw new UnknownContentTypeException("Header " + name + ": " + value);
    }
 }
```
That would cut down most of that method to lines like:

```
checkHeaderField("PF", 2, v -> v != 1);
checkHeaderField("Res", 3, v -> v == 0);
```

Reply:
But now we have Java 7 language level. It is possible to write something like this, but what we can have on code level (one line instead two) and on stack.

Reply:
Argh, shit, we’re _still_ on 7? Okay, then disregard that. :/

Reply:
Sorry for dragging this up again but I think @ArneBab recently mentioned that we can actually move to Java 8 so I’d be extremely happy if this could be the first code in Fred to use lambdas ’n shit. :)

Reply:
And I think after this I don’t have any more comments and/or requests. :) Thanks for sticking with this!